server/logging/firehose.go

```diff
@@ -0,0 +1,141 @@
+package logging
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/firehose"
+	"github.com/aws/aws-sdk-go/service/firehose/firehoseiface"
+	"github.com/pkg/errors"
+)
+
+const (
+	firehoseMaxRetries = 3
+
+	// See
+	// https://docs.aws.amazon.com/sdk-for-go/api/service/firehose/#Firehose.PutRecordBatch
+	// for documentation on limits.
+	firehoseMaxRecordsInBatch = 500
+	firehoseMaxSizeOfRecord   = 1000 * 1000     // 1,000 KB
+	firehoseMaxSizeOfBatch    = 4 * 1000 * 1000 // 4 MB
+)
+
+type firehoseLogWriter struct {
+	client firehoseiface.FirehoseAPI
+	stream string
+}
+
+func NewFirehoseLogWriter(region, id, secret, stream string) (*firehoseLogWriter, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Credentials: credentials.NewStaticCredentials(id, secret, ""),
+		Region:      &region,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "create Firehose client")
+	}
+	client := firehose.New(sess)
+
+	f := &firehoseLogWriter{client, stream}
+	if err := f.validateStream(); err != nil {
+		return nil, errors.Wrap(err, "create Firehose writer")
+	}
+	return f, nil
+}
+
+func (f *firehoseLogWriter) validateStream() error {
+	out, err := f.client.DescribeDeliveryStream(
+		&firehose.DescribeDeliveryStreamInput{
+			DeliveryStreamName: &f.stream,
+		},
+	)
+	if err != nil {
+		return errors.Wrapf(err, "describe stream %s", f.stream)
+	}
+
+	if (*(*out.DeliveryStreamDescription).DeliveryStreamStatus) != firehose.DeliveryStreamStatusActive {
+		return errors.Errorf("delivery stream %s not active", f.stream)
+	}
+
+	return nil
+}
+
+func (f *firehoseLogWriter) Write(logs []json.RawMessage) error {
```

Comment:
I found this write loop fairly hard to understand. Maybe it's a common idiom, I'm not sure. I'd add comments to clarify.

If I read it right, this is:
* trying to write at MaxRecord size
* so it looping over logs
* _If_ it `>= firehoseMaxRecordsInBatch`, then it attempts a write
* if not, it keeps looping appending logs
* writes things when the loop is over.

I imagine you're doing this, instead of iterating by N, since you also need to track firehoseMaxSizeOfBatch

Reply:
You seem to have it right. I will add some comments to try to clarify. I'm not sure there's a simpler way to write this loop given the two constraints (count and total size) we need to respect for batching the logs.